src/Data/String/CodePoints.purs

```diff
@@ -0,0 +1,225 @@
+count = _count isLead isTrail unsurrogate
+
+foreign import _count
+  :: (Int -> Boolean)
+  -> (Int -> Boolean)
+  -> (Int -> Int -> CodePoint)
+  -> (CodePoint -> Boolean)
+  -> String
+  -> Int
+
+
+drop :: Int -> String -> String
+drop n s = fromCodePointArray (Array.drop n (toCodePointArray s))
+
+
+dropWhile :: (CodePoint -> Boolean) -> String -> String
+dropWhile p s = drop (count p s) s
+
+
+fromCodePointArray :: Array CodePoint -> String
+fromCodePointArray = _fromCodePointArray singletonFallback
+
+foreign import _fromCodePointArray
+  :: (CodePoint -> String)
+  -> Array CodePoint
+  -> String
+
+
+indexOf :: String.Pattern -> String -> Maybe Int
+indexOf p s = (\i -> length (String.take i s)) <$> String.indexOf p s
+
+
+indexOf' :: String.Pattern -> Int -> String -> Maybe Int
+indexOf' p i s =
+  let s' = drop i s in
+  (\k -> i + length (String.take k s')) <$> String.indexOf p s'
+
+
+lastIndexOf :: String.Pattern -> String -> Maybe Int
+lastIndexOf p s = (\i -> length (String.take i s)) <$> String.lastIndexOf p s
+
+
+lastIndexOf' :: String.Pattern -> Int -> String -> Maybe Int
+lastIndexOf' p i s =
+  let s' = drop i s in
+  (\k -> i + length (String.take k s')) <$> String.lastIndexOf p s'
+
+
+length :: String -> Int
+length = Array.length <<< toCodePointArray
+
+
+singleton :: CodePoint -> String
+singleton = _singleton singletonFallback
+
+foreign import _singleton
+  :: (CodePoint -> String)
+  -> CodePoint
+  -> String
+
+singletonFallback :: CodePoint -> String
+singletonFallback (CodePoint cp) | cp <= 0xFFFF = fromCharCode cp
+singletonFallback (CodePoint cp) = fromCharCode lead <> fromCharCode trail
+  where
+    lead = ((cp - 0x10000) / 0x400) + 0xD800
+    trail = (cp - 0x10000) `mod` 0x400 + 0xDC00
+
+
+splitAt :: Int -> String -> Maybe { before :: String, after :: String }
+splitAt i s =
+  let cps = toCodePointArray s in
+  if i < 0 || Array.length cps < i
+    then Nothing
+    else Just {
+        before: fromCodePointArray (Array.take i cps),
+        after: fromCodePointArray (Array.drop i cps)
+      }
+
+
+take :: Int -> String -> String
+take = _take takeFallback
+
+foreign import _take :: (Int -> String -> String) -> Int -> String -> String
+
+takeFallback :: Int -> String -> String
+takeFallback n s = fromCodePointArray (Array.take n (toCodePointArray s))
+
+
+takeWhile :: (CodePoint -> Boolean) -> String -> String
+takeWhile p s = take (count p s) s
+
+
+toCodePointArray :: String -> Array CodePoint
+toCodePointArray = _toCodePointArray toCodePointArrayFallback unsafeCodePointAt0
+
+foreign import _toCodePointArray
+  :: (String -> Array CodePoint)
+  -> (String -> CodePoint)
+  -> String
+  -> Array CodePoint
+
+toCodePointArrayFallback :: String -> Array CodePoint
+toCodePointArrayFallback s = unfoldr decode (fromFoldable (Char.toCharCode <$> String.toCharArray s))
+  where
+  decode :: List Int -> Maybe (Tuple CodePoint (List Int))
+  decode (Cons h (Cons l rest)) | isLead h && isTrail l
```

Comment:
This is the only use of `lists` here. I'd like to avoid the dependency if possible. It looks like this could be defined using `zipWith` on arrays.